api/v1alpha1/zync_types.go

```diff
@@ -0,0 +1,334 @@
+	// +optional
+	ReadinessProbe *ProbeSpec `json:"readinessProbe,omitempty"`
+}
+
+// Default implements defaulting for the each zync que
+func (spec *QueSpec) Default() {
+
+	spec.HPA = InitializeHorizontalPodAutoscalerSpec(spec.HPA, zyncDefaultQueHPA)
+
+	if spec.HPA.IsDeactivated() {
+		spec.Replicas = intOrDefault(spec.Replicas, &zyncDefaultQueReplicas)
+	} else {
+		spec.Replicas = nil
+	}
+
+	spec.PDB = InitializePodDisruptionBudgetSpec(spec.PDB, zyncDefaultQuePDB)
+	spec.Resources = InitializeResourceRequirementsSpec(spec.Resources, zyncDefaultQueResources)
+	spec.LivenessProbe = InitializeProbeSpec(spec.LivenessProbe, zyncDefaultQueLivenessProbe)
+	spec.ReadinessProbe = InitializeProbeSpec(spec.ReadinessProbe, zyncDefaultQueReadinessProbe)
+}
+
+// ZyncConfig configures app behavior for Zync
+type ZyncConfig struct {
+	// Rails configuration options for zync components
+	// +operator-sdk:csv:customresourcedefinitions:type=spec
+	// +optional
+	Rails *ZyncRailsSpec `json:"rails,omitempty"`
+	// A reference to the secret holding the database DSN
+	// +operator-sdk:csv:customresourcedefinitions:type=spec
+	DatabaseDSN SecretReference `json:"databaseDSN"`
+	// A reference to the secret holding the secret-key-base
+	// +operator-sdk:csv:customresourcedefinitions:type=spec
+	SecretKeyBase SecretReference `json:"secretKeyBase"`
+	// A reference to the secret holding the zync authentication token
+	// +operator-sdk:csv:customresourcedefinitions:type=spec
+	ZyncAuthToken SecretReference `json:"zyncAuthToken"`
+	// Options for configuring Bugsnag integration
+	// +operator-sdk:csv:customresourcedefinitions:type=spec
+	// +optional
+	Bugsnag *BugsnagSpec `json:"bugsnag,omitempty"`
+}
+
+// Default sets default values for any value not specifically set in the ZyncConfig struct
+func (cfg *ZyncConfig) Default() {
+	if cfg.Rails == nil {
+		cfg.Rails = &ZyncRailsSpec{}
+	}
+	if cfg.Bugsnag == nil {
+		cfg.Bugsnag = &zyncDefaultConfigBugsnagSpec
+	}
+	cfg.Rails.Default()
+}
+
+// ZyncRailsSpec configures rails for system components
+type ZyncRailsSpec struct {
+	// Rails environment
+	// +operator-sdk:csv:customresourcedefinitions:type=spec
+	// +optional
+	Environment *string `json:"environment,omitempty"`
+	// Rails log level (debug, info, warn, error, fatal or unknown)
+	// +operator-sdk:csv:customresourcedefinitions:type=spec
+	// +kubebuilder:validation:Enum=debug;info;warn;error;fatal;unknown
+	// +optional
+	LogLevel *string `json:"logLevel,omitempty"`
+	// Rails max threads (only applies to api)
+	// +operator-sdk:csv:customresourcedefinitions:type=spec
+	// +optional
+	MaxThreads *int32 `json:"maxThreads,omitempty"`
+}
+
+// Default applies defaults for ZyncRailsSpec
+func (zrs *ZyncRailsSpec) Default() {
+	zrs.Environment = pointer.StringPtr(zyncDefaultConfigRailsEnvironment)
+	zrs.LogLevel = pointer.StringPtr(zyncDefaultConfigRailsLogLevel)
+	zrs.MaxThreads = pointer.Int32Ptr(zyncDefaultConfigRailsMaxThreads)
+}
```

Comment:
You need to use the helper functions `stringOrDefault()` and so on, otherwise the default values will always be used disregard of the user setting them or not in the custom resource.

Reply:
Yes that's true, I've seen that system is not using the helpers either
https://github.com/3scale/saas-operator/blob/main/api/v1alpha1/system_types.go#L548

Shall I fix it on this PR (I'm fixing some related system stuff also)

Reply:
Yep, that definitely needs fixing too. Let's do it straight ahead in this PR.

Reply:
fixed for zync at https://github.com/3scale/saas-operator/pull/74/commits/b79f627dadb3dd00b833c97ce0ce0cf1ba20d78f, and for system at https://github.com/3scale/saas-operator/pull/74/commits/3bb86eb50edb136e409ff9d66f5cab296e6c8c1d